src/ai/defaultai.cc

```diff
@@ -5356,22 +5352,13 @@ BuildingNecessity DefaultAI::check_building_necessity(BuildingObserver& bo,
 
 	// First we deal with training sites, they are separate category
 	if (bo.type == BuildingObserver::Type::kTrainingsite) {
-
-		if (!basic_economy_established && management_data.f_neuron_pool[17].get_position(1)) {
-			return BuildingNecessity::kNotNeeded;
-		} else if (bo.aimode_limit_status() != AiModeBuildings::kAnotherAllowed) {
-			return BuildingNecessity::kNotNeeded;
-		} else if (ts_without_trainers_ > 0 || bo.cnt_under_construction > 0 ||
-		           ts_in_const_count_ > 1) {
-			return BuildingNecessity::kNotNeeded;
-		} else if (bo.prohibited_till > gametime) {
+		if ((!basic_economy_established && management_data.f_neuron_pool[17].get_position(1)) ||
+		    bo.aimode_limit_status() != AiModeBuildings::kAnotherAllowed ||
+		    ts_without_trainers_ > 0 || bo.cnt_under_construction > 0 || ts_in_const_count_ > 1 ||
+		    bo.prohibited_till > gametime || ts_without_trainers_ > 1 ||
+		    (bo.total_count() > 0 &&
+		     soldier_trained_log.count(gametime, bo.id) / bo.total_count() < 5)) {
```

Comment:
`ts_without_trainers_ > 1` is redundant and already checked with `ts_without_trainers_ > 0`